### PR TITLE
Feature/neso particles

### PR DESCRIPTION
**Pull request by will-saunders-ukaea**
This PR:
1. Adds a NESO-Particles (NP) package following the NESO pattern of specifying a SYCL implementation.
2.  Adds a NESO-Particles dependency to NESO.

Testing:
1. Tested NP tests pass with OneAPI/IntelMPI and HipSYCL/MPICH
2. Tested NESO tests pass with OneAPI/IntelMPI and HipSYCL/MPICH (including with the NP interface branches).

Added subsection to README for NP.
Ran black on the NP package.

### CODE REVIEW

**Comment by cmacmackin** [comment]
For some broader context, why is NESO-particles being developed independently of the main NESO repository? How tightly-coupled will they be? Will you be able to build one without the other? Will NESO be dependent on a particle commit of NESO-particles or will the latter just provide an API with semantic versioning?

**Comment by will-saunders-ukaea** [comment]
NESO depends on NESO-Particles (NP) but not vice-versa. So if you want a SYCL particle framework you can install NP and your only dependencies are MPI and SYCL implementations. 

NP is currently implemented with the idea in mind that the user wants to transport particles over unstructured meshes and the user describes how the interface between the mesh and the particles works (by defining some classes and cell binning routines). This means that you can have a general mesh and as long as the shim exists between your mesh type and NP then particles can exist on your mesh. i.e. there are two existing interfaces that exist between NP and meshes 1) a built in interface between a Cartesian mesh and NP and 2) the shim between nektar 2D meshes and NP. NP can store data cell-wise but can also revert to a more classic single block of particles per MPI rank.

The tight coupling is being developed on the NESO side on the basis that the particle data is stored in NP data structures.

> Will NESO be dependent on a particle commit of NESO-particles or will the latter just provide an API with semantic versioning?
Good question - this is something we should decide.

**Comment by cmacmackin** [comment]
Thanks, that's very helpful!

I would recommend trying to establish an API with semantic versioning through which NESO will interact with NP. This will allow the dependency info in the NESO package to ensure it is being built against a suitable version of NP. We can merge this PR before you have done that however.

File: packages/neso-particles/package.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+class NesoParticles(CMakePackage):
+    """NESO-Particles"""
+
+    homepage = "https://excalibur-neptune.github.io/NESO-Particles/"
+
+    git = "https://github.com/ExCALIBUR-NEPTUNE/NESO-Particles.git"
+
+    version("0.1.0", commit="2e1b1aac6f4f9c22b31787cf300fcc2b4914cbe1", preferred=True)
+    # 22/08/2022
+    version("0.0.1-f54dc8", commit="10a54a410ddd0eb5d1debc702f415dd798eb89ee")
```
Comment (cmacmackin) [inline packages/neso-particles/package.py:15]: I'd recommend against including a hash in the version. If people want to build against a specific hash it is possible to request that when [specifying a version](https://spack.readthedocs.io/en/latest/basic_usage.html#version-specifier) for spack to fetch.

**Review by cmacmackin — CHANGES_REQUESTED** [review]
This looks good. I only have one comment for you to follow up on.

File: packages/neso-particles/package.py (hunk 1 above)
Comment (will-saunders-ukaea) [reply packages/neso-particles/package.py:15]: Specify a tag instead? (spack packaging guide seems to prefer specifying commits here https://spack.readthedocs.io/en/latest/packaging_guide.html). Not specifying a tag/commit seems fragile, i.e. I can add
```
version("0.2.0")
```
to the package and do
```
spack install neso.neso-particles@0.2.0 ^neso.hipsycl ^mpich
```
and spack happily installs something and returns a successful install. Except NESO-Particles doesn't have a 0.2.0 release. So it definitely hasn't installed 0.2.0.....

Comment (will-saunders-ukaea) [reply packages/neso-particles/package.py:15]: Oh I see, I thought you were referring to line 13 and was confused.

Comment (will-saunders-ukaea) [reply packages/neso-particles/package.py]: Pushed update.

**Review by cmacmackin — APPROVED** [review]

**Pull request merged**